### PR TITLE
misc updates and changes to the embedded UI

### PR DESCRIPTION
**Pull request by devoncarew**
misc updates and changes to the embedded UI:

- implement code completion
- add a 'format' button
- rename some of the existing buttons (`Test` ==> `Test code`, `Console` ==> `Output`, `Check code` ==> `Run`, `Reset code` ==> `Reset`)
- adjust the in-line error dialog; it now shows all the errors, not a count summary, has a max height, and supports scrolling within the dialog if necessary
- clear more toasts (the execution one one analysis results, and vice versa)
- adjust the analysis timings so we don't request analysis while the user is in the middle of typing
- make the test tab read-only
- ensure we analyze the same code we're compiling (otherwise we get unused imports warnings)
- fix a layout issue when the user had long source lines

Screen Shot 2019-04-12 at 2 28 52 PM

Screen Shot 2019-04-12 at 5 59 13 PM

### CODE REVIEW

File: lib/experimental/new_embed.dart [hunk 1]
```diff
@@ -245,81 +268,110 @@ class NewEmbed {
...
+    SourceRequest input = SourceRequest()..source = originalSource;
+
+    try {
+      formatButton.disabled = true;
+      FormatResponse result =
+          await dartServices.format(input).timeout(serviceCallTimeout);
+
+      formatButton.disabled = false;
+
+      if (originalSource != result.newString) {
```
Comment (domesticmouse) [inline lib/experimental/new_embed.dart]: Should we abort if the user has changed their source code since hitting the format button?

File: lib/services/execution_iframe.dart [hunk 2]
```diff
@@ -133,7 +133,8 @@ window.onerror = function(message, url, lineNumber, colno, error) {
     if (modulesBaseUrl != null) {
       requireConfig = '''
 require.config({
-  "baseUrl": "$modulesBaseUrl"
+  "baseUrl": "$modulesBaseUrl",
+  waitSeconds: 60
```
Comment (domesticmouse) [inline lib/services/execution_iframe.dart]: I'm guessing this content is inline JSON, so shouldn't `waitSeconds` be `"waitSeconds"`?

Comment (devoncarew) [reply lib/services/execution_iframe.dart]: Yup, it is; I'll update this.

The default timeout for require.js is 7 seconds - I was hitting that when using a hotspot connection via my phone.

File: lib/experimental/new_embed.dart (hunk 1 above)
Comment (devoncarew) [reply lib/experimental/new_embed.dart]: Yup, good point - will update this.

File: lib/experimental/new_embed.dart [hunk 3]
```diff
@@ -105,12 +112,12 @@ class NewEmbed {
           ..theme = 'elegant'
           ..mode = 'dart'
           ..showLineNumbers = true;
-
     userCodeEditor.document.onChange.listen(_performAnalysis);
 
     testEditor = editorFactory.createFromElement(querySelector('#test-editor'))
       ..theme = 'elegant'
       ..mode = 'dart'
+      ..readOnly = true
```
Comment (RedBrogdon) [inline lib/experimental/new_embed.dart]: We've been leaving the test editor writeable in the short-term as a convenience so that Kathy (and now Jon) can use the live UI to write/test gists as they're drafting their codelabs.

File: lib/experimental/new_embed.dart [hunk 4]
```diff
@@ -245,81 +268,114 @@ class NewEmbed {
...
       final dartServices = deps[DartservicesApi] as DartservicesApi;
-      final input = SourceRequest()..source = userCodeEditor.document.value;
-      final lines = Lines(input.source);
-      final request = dartServices.analyze(input).timeout(serviceCallTimeout);
+      final userSource = context.dartSource;
+      // Create a synthesis of the user code and other code to analyze.
+      final fullSource = '$userSource\n'
+          '${context.testMethod}\n'
+          '${executionSvc.testResultDecoration}\n'
+          'var resultFunction = _result;\n';
```
Comment (RedBrogdon) [inline lib/experimental/new_embed.dart]: What is this line doing? I don't see the string "resultFunction" anywhere else in the PR, so I thought I'd ask.

File: lib/experimental/new_embed.dart [hunk 5]
```diff
@@ -245,81 +268,114 @@ class NewEmbed {
...
+      String message = issue.message;
+      if (message.endsWith('.')) {
+        message = message.substring(0, message.length - 1);
       }
+      return '$message - line ${issue.line}';
+    }).toList();
 
-      analysisResultBox.showStrings([message], style);
-    }
+    analysisResultBox.showStrings(messages, FlashBoxStyle.warn);
```
Comment (RedBrogdon) [inline lib/experimental/new_embed.dart]: There's also a showElements if you'd like to make them anchors that fire playground's jumpTo when clicked. I could also do that myself in a separate PR.

File: web/experimental/new_embed.css [hunk 6]
```diff
@@ -82,6 +97,10 @@ body {
   border-left: 1px solid #e1e4e8;
 }
 
+#user-code-editor.web-output-showing {
+  width: calc(100% - 260px);
```
Comment (RedBrogdon) [inline web/experimental/new_embed.css]: Way out of scope here, but I've been wondering if we'll eventually need to let the gist or a query string parameter determine the width of the web output view. Some codelab authors might need to guarantee a certain number of pixels.

File: lib/experimental/new_embed.dart (hunk 3 above)
Comment (devoncarew) [reply lib/experimental/new_embed.dart]: I'll remove this (and add a todo to make it read-only later).

File: lib/experimental/new_embed.dart (hunk 4 above)
Comment (devoncarew) [reply lib/experimental/new_embed.dart]: Without this, the `_result` function can cause a 'method not used' analysis issue to show. This ensures that we have at least one use of it. I'll make a comment.

File: lib/experimental/new_embed.dart (hunk 5 above)
Comment (devoncarew) [reply lib/experimental/new_embed.dart]: If you don't mind updating after landing, that would be great.

File: web/experimental/new_embed.css (hunk 6 above)
Comment (devoncarew) [reply web/experimental/new_embed.css]: We could do as a query param. There's also some value is just having fewer options, and encouraging authors to be creative within those constraints.

File: lib/experimental/new_embed.dart (hunk 4 above)
Comment (RedBrogdon) [reply lib/experimental/new_embed.dart]: Got it. Thanks.

File: lib/experimental/new_embed.dart (hunk 5 above)
Comment (RedBrogdon) [reply lib/experimental/new_embed.dart]: SGTM

**Comment by devoncarew** [comment]
landing...
